### PR TITLE
Allow delegates to create pending competitions

### PR DESCRIPTION
**Pull request by jfly**
@timhabermaas, would you kindly take a look? Thanks!

### CODE REVIEW

File: WcaOnRails/app/controllers/competitions_controller.rb [hunk 1]
```diff
@@ -48,6 +51,9 @@ def create
         @competition.errors[:competition_id_to_clone] = "invalid"
       end
     end
+    if current_user.any_kind_of_delegate?
+      @competition.delegates << current_user
+    end
 
     if @competition.errors.size == 0 && @competition.save
```
Comment (timhabermaas) [inline WcaOnRails/app/controllers/competitions_controller.rb]: If you don't call `valid?` (or `save`) on `@competition` first, `@competition.errors.size` will always be `0`.

File: WcaOnRails/app/controllers/competitions_controller.rb [hunk 2]
```diff
@@ -19,28 +27,23 @@ class CompetitionsController < ApplicationController
...
     else
       # Cloning an existing competition!
       competition_to_clone = Competition.find_by_id(new_competition_params[:competition_id_to_clone])
       if competition_to_clone
-        # Don't clone the showAtAll or isConfirmed bits.
-        @competition = Competition.new(competition_to_clone.as_json.merge(new_competition_params).merge(showAtAll: false, isConfirmed: false))
+        # Don't clone showAtAll, isConfirmed, id, name, and cellName.
+        @competition = Competition.new(competition_to_clone.as_json.merge("showAtAll" => false, "isConfirmed" => false, "id" => nil, "name" => nil, "cellName" => nil).merge(new_competition_params))
         @competition.organizers = competition_to_clone.organizers
         @competition.delegates = competition_to_clone.delegates
```
Comment (timhabermaas) [inline WcaOnRails/app/controllers/competitions_controller.rb]: I'd move this stuff to `Competition`: `@competition.clone_from(other_competition)`. No need for the controller to know about all the columns a competition table has.

File: WcaOnRails/app/mailers/competitions_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class CompetitionsMailer < ApplicationMailer
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.competitions_mailer.notify_board_of_confirmed_competition.subject
+  #
```
Comment (timhabermaas) [inline WcaOnRails/app/mailers/competitions_mailer.rb]: This works even though you've set the `subject` explicitly in line 13?

File: WcaOnRails/app/models/competition.rb [hunk 4]
```diff
@@ -35,6 +35,33 @@ def cleanup_event_specs
     self.eventSpecs ||= ""
   end
 
+  before_validation :create_id_and_cell_name
```
Comment (timhabermaas) [inline WcaOnRails/app/models/competition.rb]: Could you add one or two test cases for this to the `Competition` model?
```ruby
c = Competition.new(name: "foo 2015")
c.valid?
expect(c.id).to eq "foo-2015"
```

File: WcaOnRails/app/controllers/competitions_controller.rb (hunk 1 above)
Comment (jfly) [inline WcaOnRails/app/controllers/competitions_controller.rb]: This is "if we haven't run into any problems *yet*, lets try to save. See line 51. Perhaps it's possible to move this cloning logic down into the model more?

File: WcaOnRails/app/mailers/competitions_mailer.rb (hunk 3 above)
Comment (jfly) [inline WcaOnRails/app/mailers/competitions_mailer.rb]: Stale comment from the scaffolding! I will nuke the comment.

File: WcaOnRails/app/models/competition.rb (hunk 4 above)
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: There's actually a test just like this in the competition controller spec. See https://github.com/jfly/worldcubeassociation.org/blob/a5eca69ea816ba5694c1f040aae76f5415743c03/WcaOnRails/spec/controllers/competitions_controller_spec.rb#L45-L50. I'll add a test to the competition_spec as well, though. Doesn't cost us anything.

File: WcaOnRails/app/controllers/competitions_controller.rb (hunk 1 above)
Comment (jfly) [inline WcaOnRails/app/controllers/competitions_controller.rb]: Welp, turns out if you remove this `@competition.errors.size == 0` from the `if` condition, all our tests still pass =(. I'll add a test for that first, and then see if I can refactor this to get rid of it.

Comment (timhabermaas) [inline WcaOnRails/app/controllers/competitions_controller.rb]: Oh, totally missed this. :sob: 

Moving it to the competition model sounds reasonable. I'm not sure how right now, though.

Comment (jfly) [inline WcaOnRails/app/controllers/competitions_controller.rb]: Got it! I was able to push this down into the Competition model in
https://github.com/jfly/worldcubeassociation.org/commit/680a4f1f57636e17788654cd546b3e6c6726bb35
.

**Comment by timhabermaas** [comment]
LGTM except for the commit history. Why are my commits from `master` in here?

Also: I wouldn't have moved the cloning to an `after_validation` callback, but would've preferred an explicit method for cloning. The `competition_id_to_clone` attribute on the Competition seems to be only there to make a control flow decision later in the life cycle of a competition, but it isn't really part of the state of an actual competition. Callbacks are also hard to debug and reason about. What about `if params[:competition_id_to_clone].present? then Competition.find_by_id(...).clone!` in the controller?

**Comment by jfly** [comment]
I'm not sure what happened with the commit history. I'm in a habit of running `git pull --rebase`, and things went pretty poorly for some reason that I didn't understand.

Hmm... I want to provide an error message to the user when they provide an invalid competition_id_to_clone, which I think means it needs to be part of the Competition model, so we can [add errors to it](https://github.com/cubing/worldcubeassociation.org/pull/176/files#diff-848ffc153c457b964c6c8a9d204cea61R47).

**Comment by timhabermaas** [comment]
> I'm in a habit of running git pull --rebase, and things went pretty poorly for some reason that I didn't understand.

Hmm, this *should* work if you use it just for this branch. But that's probably not what you did since there are commits from `master` in here. I'd just `rebase -i`, clean up the commits and force push at this point.

> which I think means it needs to be part of the Competition model

Technically not, but that's how Rails wants you to do it. Sad reality. So, leave it as it is. We can still refactor it if it bites us. E.g. This behavior is weird:

```ruby
c = Competition.new(competition_id_to_clone: 1)
c.venue # => nil
c.valid?
c.venue # => "Foo City Hall"
c.competition_id_to_clone = 2
c.venue # => "Foo City Hall"
```

You could probably also ignore the errors which come from `competition_id_to_clone` by either
* Pre-filling the fields on the client if the user chooses a competition to clone and not even bothering with the attribute on the server side
* Using a select box (with some fuzzy searching) for the competition id in order to prevent people from submitting non-existent `id`s

But I would consider that nice-to-have-stuff for the future.

**Comment by jfly** [comment]
I think the source of all my troubles was trying to pull in the commits from @neodude's branch. Just spent a couple hours playing around with git, and I think I got things straightened out =).

For the record, the code *is* using twitter's typeahead to autocomplete competition ids for users.

**Pull request merged**